app/presenters/documents_info_export_presenter.rb

```diff
@@ -0,0 +1,86 @@
+class DocumentsInfoExportPresenter
+  attr_reader :lead_organisation, :type
+
+  def initialize(lead_organisation:, type:)
+    @lead_organisation = lead_organisation
+    @type = type
+  end
+
+  def call
+    document_ids.sort.map do |id|
+      { document_id: id,
+        document_information: {
+          locales: locales(id),
+          subtypes: subtypes(id),
+          lead_organisations: lead_orgs(id)
+        } }
+    end
+  end
+
+  private
+
+  def document_ids
+    @document_ids ||= type.constantize
+      .joins(:organisations)
+      .where(edition_organisations: { organisation_id: organisation_id, lead: true })
+      .where(type: type)
+      .latest_edition
+      .pluck(:document_id)
+  end
+
+  def organisation_id
+    Organisation.find_by(content_id: lead_organisation).id
+  end
+
+  def locales(doc_id)
+    locales_hash[doc_id].split(',')
+  end
+
+  def subtypes(doc_id)
+    if type_has_subtype?
+      subtypes_hash[doc_id].split(',').map do |subtype_id|
+        (type + 'Type').constantize.find_by_id(subtype_id.to_i).singular_name
+      end
+    end
```

Comment:
This seems to be the bit where we have the greatest intersection between the type input. I'm wondering if we can avoid needing it by just treating everything as having any one of the types and expect that the query sorts this out for us - since we already query in that way.

So something along the lines of:

```
Edition
      .where(document_id: document_ids)
      .group(:document_id)
      .pluck(:document_id,
             "GROUP_CONCAT(DISTINCT(editions.news_article_type_id))",
             "GROUP_CONCAT(DISTINCT(editions.speech_type_id))",
             "GROUP_CONCAT(DISTINCT(editions.publication_type_id))",
             "GROUP_CONCAT(DISTINCT(editions.corporate_information_page_type_id))")
      .each_with_object do |(document_id, news, speeches, publications, corporate), memo|
         news_article_types = news.split(",").map { |id| NewsArticleType.find_by_id(id)&.key }
         speech_types = speech.split(",").map { |id| SpeechType.find_by_id(id)&.key }
         ...
         memo[document_id] = [news_article_types, speeches_types, publications_types, corporate_types].flatten
      end
```

Though I appreciate my suggestion here is pretty heavy in a method chain so could be broken up a bit.

Also couple of things I've suggested here: use key rather than singular name - seems to be consistent with https://github.com/alphagov/whitehall/pull/4963, and we should consider that the type class doesn't exist (example below)

```
[21] pry(main)> Publication.where.not(publication_type_id: PublicationType.all.map(&:id)).count
Creating scope :open. Overwriting existing method Consultation.open.
   (57.1ms)  SELECT COUNT(*) FROM `editions` WHERE `editions`.`type` IN ('Publication') AND (`editions`.`state` != 'deleted') AND (`editions`.`publication_type_id` NOT IN (1, 2, 3, 19, 4, 5, 15, 6, 7, 17, 10, 12, 14, 18, 20, 21, 8, 13, 22, 999, 1000))
=> 6
```

Reply:
Done, have split the method up as suggested.